### PR TITLE
[Music]Filter music items by media source

### PR DESCRIPTION
**Pull request by DaveTBlake**
Add music library facility to filter artists, albums and songs by media source.

This is particularly useful for those with large and diverse music collections, and brings some of the things that advanced users could sometimes achieve using custom nodes with path rules within reach of every user. It also could encourage more users to use the music library rather than just access their music using fileview mode.

Categorising music by putting it under separate folders is a very natural thing to do, and with the introduction of the Artist Information Folder in v18, there is no longer any need to try to have all the music by an artist, and only that artist, under one unique folder in order to have local art and artist.nfo files.  Hence users can arrange their music files into whatever folders they like e.g. "Opera", "Kids stuff", "Party Music", "Top Jazz Music", "Why did I buy this" etc. and then easily, via the media source dialog, add these as separate sources.

### CODE REVIEW

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -5770,129 +5940,728 @@ bool CMusicDatabase::GetArtistFolderName(const std::string &strArtist, const std
...
+    // cleanup
+    m_pDS->close();
+
+    return true;
+  }
+  catch (...)
+  {
+    CLog::Log(LOGERROR, "%s failed", __FUNCTION__);
+  }
+}
```
Comment (Rechi) [inline xbmc/music/MusicDatabase.cpp]: return missing

**Comment by zag2me** [comment]
Tested using the build you posted on the forum thread and it works as expected on my basic source list. I also checked the database and the new source tables populate fine.

**Comment by DaveTBlake** [comment]
This facility has been tested by users and received positive feedback, and I have given it a thorough test on both SQLite and MySQL allowing for many combinations of music source adding and removal. 

Although it is a new feature it is low risk, and I would like to get it into master soonest. I doubt if any team members have either time or interest to look deeply at it, and there is nothing to gain by just waiting.  Any objections to my merging it?

**Comment by Rechi** [comment]
jenkins build this please

**Comment by phate89** [comment]
Great feature. Video side needs this too.. One question @DaveTBlake , did you consider to actually move the sources to the db instead of syncing them? what do you think about it?

**Comment by DaveTBlake** [comment]
Thanks @phate89, I left the commits separate to show the separate steps and hopefully help anyone that wants to do the similar for video. :)

As for replacing the sources.xml file completely, that is attractive in many ways, the xml file and source add/remove is so lose at the moment. But it was a more significant change than I could do and get it into v18, and I so wanted to show what could be done with source filtering.  But maybe we talk about it for v19 eh?

**Comment by phate89** [comment]
I agree that source remove/add with xml is loose, also a lot of the code added is just to sync the changes.. the main problem is that in that xml file there are other types of sources and it's not simply needed to move video and music source. anyway it's still a good idea to do something about it forthe next version..
I'll look into it replicating to video library when I have time

File: xbmc/music/windows/GUIWindowMusicBase.h [hunk 2]
```diff
@@ -53,10 +53,13 @@ class CGUIWindowMusicBase : public CGUIMediaWindow, public IBackgroundLoaderObse
   void DoScan(const std::string &strPath, bool bRescan = false);
   void RefreshContent(const std::string& strContent);
 
-  /*! \brief Prompt the user if he wants to start a scan for this folder
-  \param path the path to assign content for
+  /*! \brief Once a music source is added, store source in library, and prompt
+  the user to scan this folder into the library
+  \param strPath the music source path
+  \param strName the name of the music source
```
Comment (phate89) [inline xbmc/music/windows/GUIWindowMusicBase.h]: these params are different than the one in the function

File: xbmc/music/MusicDatabase.cpp [hunk 3]
```diff
@@ -5770,115 +5940,696 @@ bool CMusicDatabase::GetArtistFolderName(const std::string &strArtist, const std
...
+
+      // Find albums by song path, building WHERE for multiple source paths
+      std::vector<int> albumIds;
+      Filter extFilter;
+      strSQL = "SELECT DISTINCT idAlbum FROM song ";
+      extFilter.AppendJoin("JOIN path ON song.idPath = path.idPath");
+      for (const auto& path : vecPaths)
+        extFilter.AppendWhere(PrepareSQL("path.strPath LIKE '%s%%%%'", path.c_str()), false);
+      if (!BuildSQL(strSQL, extFilter, strSQL))
+        return false;
```
Comment (phate89) [inline xbmc/music/MusicDatabase.cpp]: only minor stuff but it's an int function you should return -1?

**Comment by Uatschitchun** [comment]
That's a really nice one!! :-)

**Pull request merged**